### DISCUSSION TITLE
Support email signup universal link redirect

### DISCUSSION THREAD

**Discussion by percarlsen** [supabase/supabase#25923] [Feature Requests] [answered]
I'm building a flutter app primarily for iOS (for now). The app has email signup, and I want to redirect the user to a certain screen after pressing the link in the confirmation email. To do so, I'm configuring the `SITE_URL` in supabase to point to the page I want to direct the user to. The issue is that my app uses universal links (as per [Apple's recommendation](https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app)), but the `SITE_URL` only supports custom schemes, not universal links ([documentation](https://supabase.com/docs/guides/auth/native-mobile-deep-linking?queryGroups=platform&platform=flutter&queryGroups=os&os=ios)).

I've worked around this for now by using custom scheme for this purpose only, but I'd prefer to only use universal link. Would it be possible to support universal links in `SITE_URL`?

**Comment by dshukertjr** [comment]
Universal link as in a regular URL for a website, correct? You can certainly add regular URLs for a website into `SITE_URL` or the `Redirect URLs` in your Supabase dashboard. Supabase fully supports both custom schemes and universal links.

**Reply by percarlsen** [reply]
Thanks for answering! As in a regular URL for a website, yes :-) In that case, the documentation ([link](https://supabase.com/docs/guides/auth/native-mobile-deep-linking?queryGroups=platform&platform=flutter&queryGroups=os&os=ios)) might be wrong?
> The redirect callback url should have this format [YOUR_SCHEME]://[YOUR_HOSTNAME]. Here, io.supabase.flutterquickstart://login-callback is just an example, you can choose whatever you would like for YOUR_SCHEME and YOUR_HOSTNAME as long as the scheme is unique across the user's device. For this reason, typically a reverse domain of your website is used.

I am using deep link with regular URL for all other purposes in the same app, and it works just fine. E.g. if I press the url https://my-domain.com/foo, the user is directed to the /foo page in the app, indicating that the deep linking is configured properly. However when configuring the redirect with email signup, it doesn't work when I set `SITE_URL` to that regular URL (I have tried various variations with/without trailing slash, with/without https, http, www etc). The link is just opened in the browser instead of opening the app. If I set `SITE_URL` to com.my-domain://foo, it works fine - I'd just like to avoid doing this to stay in line with Apple's recommendation.

**Reply by dshukertjr** [reply]
@percarlsen Hmm, that is weird. If you are taken to the correct link upon clicking the sign-up confirmation link, then everything should be set up properly on the Supabase's end...

**Reply by percarlsen** [reply]
Something could of course be wrong with the setup on my end, but I don't really know how to further debug this. It does seem a bit unlikely though as I've already confirmed that normal URL deep links works fine in other contexts. I can give it another go later this week, maybe I've missed something.

**Reply by Vedsaga** [reply, marked as answer]
The same issues, we were dealing few days back, at the end came to know that it's not issue with the `Supabase`. Instead it is the expected behaviour, this is how it should work.  

# Case 1 - User click on the link
If user click on the link then OS get notified intent to process the `uri` on which user clicked. And so now OS look through the registered links ( which apps intended to consume) if any App is found which want to consume link ( based on info.plist config) then app hand over the link to that app. From there we continue processing the link in our app normally.

# Case 2 - User manually type/copy-paste link in browser
If user manually type/copy-paste standard link (`http` or `https`) in browser, then browser does let the intent reach to the OS. Due that OS never gets chance to hand-over the link to the respective app. 

It is well-known issues in Chromium browsers which is marked as `Won't Fix`, and it seems Firefox also adopted same strategy so far as we tested in our app. Not sure about the Safari or any other browsers. _Have below shared the links_

# Case 3 Link Redirect
As url-config in Supabase if we specify the redirect-link as normal `http` or `https` which is classified as `app-links` in iOS, then as soon Supabase redirect to `http` or `https`, browser does not let intent to reach the OS, which is app never opens.  As mentioned in the `case 2` just above.

However is `uri` does not contain typical `http` or `https` schema, then browser hand-over the URI to OS, which then hand-over the Link to respective app.

### Links References
 
`Stackoverflow Feb 11, 2020` -> [Deep linking working with adb, not in browser](https://stackoverflow.com/questions/60165789/deep-linking-working-with-adb-not-in-browser)

`Chromium Issue May 6, 2015` -> [ Allow generation of a link to go to fallback URL when typed in and go to market when clicked](https://issues.chromium.org/issues/41172585)

`Github  Sep 5, 2017` -> [[Android] Deep Linking not working from Browser](https://github.com/react-navigation/react-navigation/issues/2525)

`Stackoverflow May 6, 2016` -> [Deep linking not working in chrome](https://stackoverflow.com/questions/37069816/deep-linking-not-working-in-chrome)

Hope this helps...

cc:
@percarlsen

**Reply by percarlsen** [reply]
Thanks for the detailed description @Vedsaga! I have Brave browser, which uses Chromium, on my test device, so that probably explains why I can't get it to work. But if I understand correctly you suggest that it would work if Safari was the default browser? I'll give that a go.

Update: confirming that redirecting to a regular URL, e.g. https://my-domain.com/foo, works fine when Safari is the default browser on the device.

**Reply by Vedsaga** [reply]
It seems even in Safari they have some what similar behaviour to Chromium...

[How to get a universal link to open the app instead of safari?](https://stackoverflow.com/questions/43843093/how-to-get-a-universal-link-to-open-the-app-instead-of-safari)

cc:
@percarlsen

**Reply by percarlsen** [reply]
@Vedsaga just updated my message above confirming that I'm able to make it work when safari is the default browser. The SO thread you share is pretty old and might not be relevant anymore.

**Reply by Vedsaga** [reply]
That's great news, yaa that was 8 year old thread...

**Reply by percarlsen** [reply]
Yes, at least that explains why I wasn't able to make it work. It seems like custom schema is the way to go to ensure it works regardless of which browser the user has, unfortunately.

**Reply by Vedsaga** [reply]
> Yes, at least that explains why I wasn't able to make it work. It seems like custom schema is the way to go to ensure it works regardless of which browser the user has, unfortunately.

There is one may is to add button which user can click and get back, if you don't have any content on original website...

For now in our project we are only using Custom-Schema for Auth related flow... remaining we rely on `http`-`https` only... so that in future if app and user click on the link then app open else if user manually copy-paste the the link then browser will handle...

cc:
@percarlsen

**Reply by percarlsen** [reply]
Yes, that's an option. Another possible alternative could be to:
- In the email; redirect the user to the app with `{{ .RedirectTo }}`, and pass `{{ .ConfirmationURL }}` as a parameter (note: we're not clicking the confirmation at this point, just forwarding it to the app.
- The app immediately opens the `{{ .ConfirmationURL }}` in the background, then navigates the user to the appropriate screen once the confirmation is done.

I haven't tried it, but think it could work. However I'm in the early development stage of my app so I'll just fall back to using the custom scheme for now and potentially improve this later on

**Reply by percarlsen** [reply]
Another downside of using a custom scheme for this purpose is that if the confirmation email link is pressed from a device that doesn't have the app installed (e.g. your laptop), the browser will not understand the URL and hence just open a blank tab. However if we were using a universal link (normal URL), the browser could at least open your website which could display an appropriate message.
